### DISCUSSION TITLE
Auto Improving schema

### DISCUSSION THREAD

**Discussion by Mupli** [e3dio/packBytes#8] [Ideas]
what about autogenerating schema based on js object. With auto improving  capabilities. 

```javascript 
data  = {
title: "BLA BLA"
}

const encoder = new PB()

//Server A
const [dataEncoded, schema, version] = encoder.encode(data)


if (schema) 
send(dataEncoded,  version, schema)
else 
send(dataEncoded, version)

//  Receiver 
const encoder = new PB()
on(dataEncoded, version, schema = null) {
    const data = encoder.decode(buffer, version, schema )
}

**Comment by e3dio** [comment]
A single data object is not enough to generate a schema, you need the entire range of all possible data objects to create an efficient schema. You can create some auto gen function for your case but everyone's case is different so adding some auto gen function to the library is not good. You get the smallest size by designing schemas, not auto generating them. I should add more examples to the docs.

For schema versions my apps work by on websocket first open schema is sent as JSON string to client, client loads schema, and all future messages are decoded with that schema. If schema changes you send new schema with Json text string, so there is no need to track versions.

**Reply by e3dio** [reply]
Also looks like you are trying to create a schema for each item sent, which is bad as I explained here https://github.com/e3dio/packBytes/discussions/5 no point creating a schema for each item individually, you are doubling or more the size sent with each item, defeats the purpose of encoding. You are supposed to design 1 schema for all data sent. the `schemas` type allows multiple schemas within 1 schema which I use often

**Reply by e3dio** [reply]
I should update the docs with more use-case examples like the `schemas` type and show how you can send schema to the client via Json
